x-pack/legacy/plugins/actions/server/builtin_action_types/webhook.ts

```diff
@@ -0,0 +1,154 @@
+
+const HeadersSchema = schema.recordOf(schema.string(), schema.string());
+
+export const CompositeUrlSchema = schema.object({
+  host: schema.string(),
+  port: portSchema(),
+  path: nullableType(schema.string()),
+  scheme: schema.oneOf(
+    [schema.literal(WebhookSchemes.HTTP), schema.literal(WebhookSchemes.HTTPS)],
+    {
+      defaultValue: WebhookSchemes.HTTP,
+    }
+  ),
+});
+
+const ConfigSchema = schema.object({
+  url: schema.oneOf([schema.string(), CompositeUrlSchema]),
+  method: schema.oneOf([schema.literal(WebhookMethods.POST), schema.literal(WebhookMethods.PUT)], {
+    defaultValue: WebhookMethods.POST,
+  }),
+  headers: nullableType(HeadersSchema),
+});
+
+// secrets definition
+export type ActionTypeSecretsType = TypeOf<typeof SecretsSchema>;
+const SecretsSchema = schema.object({
+  user: schema.string(),
+  password: schema.string(),
+});
+
+// params definition
+export type ActionParamsType = TypeOf<typeof ParamsSchema>;
+const ParamsSchema = schema.object({
+  body: schema.maybe(schema.string()),
+});
+
+// action type definition
+export const actionType: ActionType = {
+  id: '.webhook',
+  name: 'webhook',
+  validate: {
+    config: ConfigSchema,
+    secrets: SecretsSchema,
+    params: ParamsSchema,
+  },
+  executor,
+};
+
+function asComposedUrl(url: string | TypeOf<typeof CompositeUrlSchema>): string {
+  if (typeof url === 'string') {
+    return url;
+  }
+  return `${url.scheme}://${url.host}${url.port ? `:${url.port}` : ''}${url.path || ''}`;
+}
+
+function describeWebhookErrorSource(error: AxiosError): string {
+  if (error.response) {
+    // The request was made and the server responded with a status code
+    // that falls out of the range of 2xx
+    return `remote webhook failure`;
+  } else if (error.request) {
+    // The request was made but no response was received
+    return `unreachable remote webhook`;
+  }
+  return `unknown webhook event`;
+}
+
+// action executor
+async function executor(execOptions: ActionTypeExecutorOptions): Promise<ActionTypeExecutorResult> {
+  const log = (msg: string) => execOptions.services.log(['warn', 'actions', 'webhook'], msg);
+
+  const id = execOptions.id;
+  const { method, url, headers = {} } = execOptions.config as ActionTypeConfigType;
+  const { user: username, password } = execOptions.secrets as ActionTypeSecretsType;
+  const { body: data } = execOptions.params as ActionParamsType;
+
+  const result: Result<AxiosResponse, AxiosError> = await promiseResult(
+    axios.request({
+      method,
+      url: asComposedUrl(url),
+      auth: {
+        username,
+        password,
+      },
+      headers,
+      data,
+    })
+  );
+
+  if (isOk(result)) {
+    const { value: response } = result;
+    log(`response from ${id} webhook event: ${response.status}`);
+
+    if (response.status >= 200 && response.status < 300) {
+      return {
+        status: 'ok',
+        data: response.data,
+      };
+    } else {
```

Comment:
Cool, I wasn't aware of this ability.

I've added this and unified the way Slack and Webhooks implement this.
This is mostly a refactoring, moving code around, but you'll note a small change in the Slack client which is that if you get a 429 we'll now retry with the default instead of dropping through to the generic error.
This feels right to me as a Too Many Requests is a client issue not server and so, we should encourage retrying after a short interval.
If the server is overloaded it'll likely respond with a 503/504, wouldn't you agree?

Reply:
If a server is overloaded, it's LB will return a 50x responses, usually :-).  Not sure how much consistency there is for particular ones, I think my test is usually just >= 500